### PR TITLE
reduce piled goroutines

### PR DESCRIPTION
**Pull request by joesis**
It's by no means a solution for https://github.com/getlantern/lantern-internal/issues/1462, but the piled up goroutines disappear with this code.

### CODE REVIEW

**Comment by myleshorton** [comment]
Keeping around for now.

**Comment by joesis** [comment]
Thanks, @oxtoacart. I should be able to revisit it recently consider that private line is relatively stable now.

File: lampshade_test.go [hunk 1]
```diff
@@ -390,6 +390,22 @@ func echoServerAndDialer(maxStreamsPerConn uint16) (net.Listener, Dialer, DialFN
...
+	}
+	defer l.Close()
+
+	conn, err := dial()
+	if !assert.NoError(t, err) {
+		return
+	}
+	conn.(Stream).Session().Close()
+	// Simply make sure it doesn't block
+	conn.Close()
```
Comment (joesis) [inline lampshade_test.go]: In master branch, this line will block forever.

File: session.go [hunk 2]
```diff
@@ -507,4 +498,17 @@ func (s *session) EMARTT() time.Duration {
...
+type sessionWriter struct {
+	s *session
+}
+
+func (w sessionWriter) Write(b []byte) (int, error) {
+	select {
+	case <-w.s.closeCh:
+		return 0, errorAlreadyClosed
+	case w.s.out <- b:
+		return len(b), nil
```
Comment (joesis) [inline session.go]: This is the main change: Stop writing if the session is closed.

**Comment by joesis** [comment]
Ok, the tests are passing now. This should be ready to review @oxtoacart.

File: sendbuffer.go [hunk 3]
```diff
@@ -70,14 +71,14 @@ func (buf *sendBuffer) sendLoop(out chan []byte) {
 				select {
 				case <-windowAvailable:
 					// send allowed
-					out <- append(frame, buf.defaultHeader...)
+					w.Write(append(frame, buf.defaultHeader...))
```
Comment (joesis) [inline sendbuffer.go]: Return values are not checked in all callers, but I'm not sure if handling the error is any better than just ignoring it.

File: session.go (hunk 2 above)
Comment (oxtoacart) [reply session.go]: I like the way you encapsulated this as a `Writer`!

**Comment by oxtoacart** [comment]
LGTM!

**Comment by joesis** [comment]
Thanks @oxtoacart ! I'll create a PR to update flashlight.

**Comment by oxtoacart** [comment]
@joesis I'm already working on updating deps :) I've got some proxybench stuff I want to pull in too.

**Comment by joesis** [comment]
Oh cool thanks! I just saw the branch.

**Pull request merged**